test/signalrclient-e2e-tests/connection_tests.cpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#pragma once
+
+#include "stdafx.h"
+#include <string>
+#include "cpprest\details\basic_types.h"
+#include "cpprest\json.h" 
+#include "connection.h"
+#include "hub_connection.h"
+#include "url.h"
+
+using namespace signalr;
+using namespace utility;
+using namespace std;
+using namespace pplx;
+
+TEST(connection_tests, send_message)
+{
+    auto conn = make_shared<connection>(url + U("raw-connection"));
+    auto message = make_shared<string_t>();
+    auto test_event = make_shared<pplx::event>();
+
+    conn->set_message_received([message, test_event](const string_t& payload)
+    {
+        *message = payload;
+        test_event->set();
+    });
+
+    conn->start().then([conn](task<void> start_task)
+    {
+        try
+        {
+            start_task.get();
+            ASSERT_EQ(conn->get_connection_state(), connection_state::connected);
+
+            web::json::value obj;
+            obj[U("type")] = web::json::value::number(0);
+            obj[U("value")] = web::json::value::string(U("test"));
+
+            conn->send(obj.serialize());
+        }
+        catch (...)
+        { }
+    });
+
+    ASSERT_FALSE(test_event->wait(10000));
+
+    ASSERT_EQ(*message, U("\"Send: test\""));
+}
+
+TEST(connection_tests, send_message_after_connection_restart)
+{
+    auto conn = make_shared<connection>(url + U("raw-connection"));
+    auto message = make_shared<string_t>();
+    auto test_event = make_shared<pplx::event>();
+
+    conn->set_message_received([message, test_event](const string_t& payload)
+    {
+        *message = payload;
+        test_event->set();
+    });
+
+    conn->start().then([conn](task<void> start_task)
+    {
+        try
+        {
+            start_task.get();
+            ASSERT_EQ(conn->get_connection_state(), connection_state::connected);
+        }
+        catch (...)
+        { }
+    }).then([]()
+    {
+        return;
```

Comment:
Empty then is not the way to go.

From: Xiaohong Tang [mailto:notifications@github.com]
Sent: Friday, March 13, 2015 11:02 AM
To: aspnet/SignalR-Client-Cpp
Cc: Pawel Kadluczka
Subject: Re: [SignalR-Client-Cpp] add c++ end-to-end tests (#109)


In test/signalrclient-e2e-tests/connection_tests.cpp<https://github.com/aspnet/SignalR-Client-Cpp/pull/109#discussion_r26408597>:

> +        *message = payload;

> +        test_event->set();

> +    });

> +

> +    conn->start().then([conn](task<void> start_task)

> +    {

> +        try

> +        {

> +            start_task.get();

> +            ASSERT_EQ(conn->get_connection_state(), connection_state::connected);

> +        }

> +        catch (...)

> +        { }

> +    }).then([]()

> +    {

> +        return;

it is different with .then and without .then method here because concurrency conn->start().then() and conn->stop().then() and conn->start().then(), here make connection in correct state stable, however looks like it is better to call .get method here since in it already catch all exception and it will be more stable.

—
Reply to this email directly or view it on GitHub<https://github.com/aspnet/SignalR-Client-Cpp/pull/109/files#r26408597>.